app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+            if (directionKeysPressed.contains(Direction.DOWN)) verticalVelocity += velocity
+            var horizontalVelocity = 0f
+            if (directionKeysPressed.contains(Direction.LEFT)) horizontalVelocity -= velocity
+            if (directionKeysPressed.contains(Direction.RIGHT)) horizontalVelocity += velocity
+            oldPos.x += horizontalVelocity
+            oldPos.y += verticalVelocity
+            oldPos.x = oldPos.x.coerceIn(0f, screenBounds.x)
+            oldPos.y = oldPos.y.coerceIn(0f, screenBounds.y)
+        }
+
+        updatePosition()
+
+        return velocity
+    }
+
+    /**
+     * If the cursor is near the edge of the screen, this will return that
+     * direction.  If not, it will return null.
+     *
+     * When in a corner, only UP or DOWN will be returned.
+     */
+    fun getEdgeOfScreenNearCursor(): Direction? {
+        val screenBounds = screenBounds ?: return null
+        return when {
+            lastKnownCursorPos.y < EDGE_OF_SCREEN_MARGIN -> Direction.UP
+            lastKnownCursorPos.y > screenBounds.y - EDGE_OF_SCREEN_MARGIN -> Direction.DOWN
+            lastKnownCursorPos.x < EDGE_OF_SCREEN_MARGIN -> Direction.LEFT
+            lastKnownCursorPos.x > screenBounds.x - EDGE_OF_SCREEN_MARGIN -> Direction.RIGHT
+            else -> null
+        }
+    }
+
+    private fun resetCursorSpeed() {
+        lastVelocity = BASE_SPEED
+        lastUpdatedAtMS = NOT_RECENTLY_UPDATED
+        directionKeysPressed.clear()
+    }
+
+    // This is taken from older code.  Crufty, but it works
+    private fun getScrollDistance(vel: Float, pos: PointF, framesPassed: Int): Pair<Int, Int>? {
+        val scrollVelReturnVal = PointF(0f, 0f)
+        val screenBounds = screenBounds ?: return 0 to 0
+
+        if (vel > 0f) {
+            val percentMaxVel = vel / MAX_VELOCITY
+            if (pos.x == 0f && directionKeysPressed.contains(Direction.LEFT)) {
+                scrollVelReturnVal.x = -percentMaxVel
+            } else if (pos.x == screenBounds.x && directionKeysPressed.contains(Direction.RIGHT)) {
+                scrollVelReturnVal.x = percentMaxVel
+            }
+
+            if (pos.y == 0f && directionKeysPressed.contains(Direction.UP)) {
+                scrollVelReturnVal.y = -percentMaxVel
+            } else if (pos.y == screenBounds.y && directionKeysPressed.contains(Direction.DOWN)) {
+                scrollVelReturnVal.y = percentMaxVel
+            }
+        }
+
+        scrollX = (scrollVelReturnVal.x * MAX_SCROLL_VELOCITY * framesPassed).toInt()
+        scrollY = (scrollVelReturnVal.y * MAX_SCROLL_VELOCITY * framesPassed).toInt()
+        return if (scrollX != 0 || scrollY != 0) {
+            scrollX to scrollY
+        } else null
+    }
+
+    @SuppressLint("CheckResult") // Does not need to be disposed as this survives for the duration of the app
+    private fun resetSpeedWhenCursorIsInactive() {
```

Comment:
nit: This is attaching an Observer but I find it confusing that it sounds like an immediate conditional state change listener (if cursor is inactive, reset the state right now). Maybe, `attachResetStateObserver`? Or even `attachStateObservers()` and we can add arbitrary subscribe code to it.